datastore/threads/semaphore.nim

```diff
@@ -0,0 +1,58 @@
+import std/atomics
+import std/locks
+
+type
+  Semaphore* = object
+    count: int
+    size: int
+    lock {.align: 64.}: Lock
+    cond: Cond
```

Comment:
I like the semaphore idea. Though wouldn't Chronos have an async semaphore mechanism?

Reply:
There is one in libp2p and I actually used it before switching to the threaded version, I think tho, for the proxy the async semaphore should be fine.